bugfix and adding test for GetFolderPath when special folder doesnt exist

maryamariyan opened the pull request:
Bugfix on GetFolderPath in Unix when special folder doesnt exist

Fixes #26677
cc: @danmosemsft

danmosemsft commented on src/System.Runtime.Extensions/src/System/Environment.Unix.cs:
```diff
@@ -18,7 +18,7 @@ public static partial class Environment
         internal static readonly bool IsMac = Interop.Sys.GetUnixName() == "OSX";
         private static Func<string, IEnumerable<string>> s_fileReadLines;
-        private static Action<string> s_directoryCreateDirectory;
+        private static Func<string,object> s_directoryCreateDirectory;
```
Spacing after comma

maryamariyan replied:
Noticed after the PR created. updating right away.

danmosemsft commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
+        public void GetFolderPath_Unix_SpecialFolderNotCreated_DoesNotThrow(Environment.SpecialFolder folder)
+        {
+            var path = Environment.GetFolderPath(folder, Environment.SpecialFolderOption.DoNotVerify);
+            Assert.False(System.IO.Directory.Exists(path));
```
Are we sure they never exist? I wouldn't assert. Just use Exists to determine whether to delete the folder at the end.

maryamariyan replied:
OK

maryamariyan commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
         }
 
+        [Theory]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]  // Tests OS-specific environment
```
All of the below InlineData passed locally. After the build is finished I intend to reduce the following InlineData. Because in order to exercise the code path that this PR fixes one of the following InlineData would be sufficient.

danmosemsft replied:
I don't care whether you have all those variations - in fact that makes it more sure that at least one will exercise the "create" path.

As you can see the logic to choose tose locations is complex, it is not clear they are all missing on all distros, and in fact your test results prove they arent:
https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L95

maryamariyan commented:
@danmosemsft thanks for your quick review. I’ll wait for the current build to finish before updating the PR with your feedback.

danmosemsft commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
+        public void GetFolderPath_Unix_SpecialFolderNotCreated_DoesNotThrow(Environment.SpecialFolder folder)
+        {
+            var path = Environment.GetFolderPath(folder, Environment.SpecialFolderOption.DoNotVerify);
+            Assert.False(System.IO.Directory.Exists(path));
```
Please avoid excessive namespace qualifications, you hae `using Systme.IO;` at the top already.

danmosemsft commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
+        [InlineData(Environment.SpecialFolder.MyPictures)]
+        [InlineData(Environment.SpecialFolder.MyVideos)]
+        [InlineData(Environment.SpecialFolder.Templates)]
+        public void GetFolderPath_Unix_SpecialFolderNotCreated_DoesNotThrow(Environment.SpecialFolder folder)
```
Since this deletes "important" folders, albeit if they don't prevoiusly exist, you might consider [Outerloop] out of paranoia.

stephentoub replied:
How does this test relate to https://github.com/maryamariyan/corefx/blob/32a11618b153175f38d38b512202de11c1e5f712/src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs#L254-L294?  If it's testing "DoNotVerify" (which I take to be the primary purpose given the "SpecialFolderNotCreated_DoesNotThrow" part of the name), do we have overlap with that other test?

stephentoub commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
+        [InlineData(Environment.SpecialFolder.Templates)]
+        public void GetFolderPath_Unix_SpecialFolderNotCreated_DoesNotThrow(Environment.SpecialFolder folder)
+        {
+            var path = Environment.GetFolderPath(folder, Environment.SpecialFolderOption.DoNotVerify);
```
var => string

stephentoub commented on src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -215,6 +215,25 @@ public void GetFolderPath_Unix_PersonalIsHomeAndUserProfile()
+            Assert.False(System.IO.Directory.Exists(path));
+            path = Environment.GetFolderPath(folder, Environment.SpecialFolderOption.Create);
+            Assert.True(System.IO.Directory.Exists(path));
+            Directory.Delete(path);
```
This seems problematic.  Presumably Environment.GetFolderPath is going to be accessing some shared location.  We might delete something that's in use, even if we only delete it if we just created it, as there could be a race condition.

maryamariyan commented:
@dotnet-bot test outerloop netcoreapp OSX x64 Debug Build

@dotnet-bot test outerloop netcoreapp OSX x64 Release Build

@dotnet-bot test outerloop netcoreapp Linux x64 Debug Build

@dotnet-bot test outerloop netcoreapp Linux x64 Release Build

maryamariyan commented:
@dotnet-bot test outerloop netcoreapp Linux x64 Debug Build

maryamariyan commented:
@dotnet-bot help

maryamariyan commented:
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Release
@dotnet-bot test outerloop netcoreapp Ubuntu16.04 Debug
@dotnet-bot test outerloop netcoreapp OSX10.12 Release
@dotnet-bot test outerloop netcoreapp OSX10.12 Debug

maryamariyan commented:
@dotnet-bot test outerloop netcoreapp CentOS7.1 Release
@dotnet-bot test outerloop netcoreapp CentOS7.1 Debug
@dotnet-bot test outerloop netcoreapp Debian8.4 Release
@dotnet-bot test outerloop netcoreapp Debian8.4 Debug
@dotnet-bot test outerloop netcoreapp Fedora24 Release
@dotnet-bot test outerloop netcoreapp Fedora24 Debug
@dotnet-bot test outerloop netcoreapp PortableLinux Release
@dotnet-bot test outerloop netcoreapp PortableLinux Debug
@dotnet-bot test outerloop netcoreapp RHEL7.2 Release
@dotnet-bot test outerloop netcoreapp RHEL7.2 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu14.04 Release
@dotnet-bot test outerloop netcoreapp Ubuntu14.04 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Release
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Debug

maryamariyan commented:
Here's a review of what the failing tests are:

The netfx failure is unrelated. Refer to issue #26770 

The outerloop failures are unrelated to this change:
- All or some of the following three System.Diagnostics.Process tests fail in outerloop
```
System.Diagnostics.Tests.ProcessTests.TestStartWithRootUserCannotElevate 
System.Diagnostics.Tests.ProcessTests.TestStartWithRootUser 
System.Diagnostics.Tests.ProcessTests.TestStartWithNormalUser 
```
- Other than the above three, only Fedora24 has the addition test failure below:
```
System.Net.Sockets.Tests.SocketOptionNameTest.MulticastInterface_Set_AnyInterface_Succeeds
```

stephentoub commented:
@maryamariyan, thanks, can you please make sure there are issues for all of those?  If there are, note these failures there, and if there aren't, open issues?

maryamariyan commented:
Thanks @stephentoub. I updated my last comment by referring to each test failure's relevant corefx issue numbers.
I added new issue #26775, the other two issues were existing.
